src/input/mod.rs

```diff
@@ -0,0 +1,490 @@
+/// source you are plan to use. See the subclasses in this module.
+#[pyclass(module = "bytewax.inputs", subclass)]
+#[pyo3(text_signature = "()")]
+pub(crate) struct InputConfig;
+
+/// Use [Kafka](https://kafka.apache.org) as the input source.
+///
+///
+/// Args:
+///
+///     brokers: List of broker addresses.
+///
+///     group_id: Group id
+///
+///     topic: Topic to which consumer will subscribe
+///
+///     batch_size: Integer defining how messages will be grouped by epoch.
+///
+/// Returns:
+///
+///     Config object. Pass this as the `input_config` argument to
+///     your execution entry point.
+#[pyclass(module = "bytewax.inputs", extends = InputConfig)]
+#[pyo3(text_signature = "(brokers, group_id, topics)")]
+pub(crate) struct KafkaInputConfig {
+    #[pyo3(get)]
+    pub brokers: String,
+    #[pyo3(get)]
+    pub group_id: String,
+    #[pyo3(get)]
+    pub topics: String,
+    #[pyo3(get)]
+    pub batch_size: u64,
+}
+
+#[pymethods]
+impl KafkaInputConfig {
+    #[new]
+    #[args(brokers, group_id, topics, batch_size = 1)]
+    fn new(
+        brokers: String,
+        group_id: String,
+        topics: String,
+        batch_size: u64,
+    ) -> (Self, InputConfig) {
+        (
+            Self {
+                brokers,
+                group_id,
+                topics,
+                batch_size,
+            },
+            InputConfig {},
+        )
+    }
+
+    fn __getstate__(&self) -> (&str, String, String, String, u64) {
+        (
+            "KafkaInputConfig",
+            self.brokers.clone(),
+            self.group_id.clone(),
+            self.topics.clone(),
+            self.batch_size,
+        )
+    }
+
+    /// Egregious hack see [`SqliteRecoveryConfig::__getnewargs__`].
+    fn __getnewargs__(&self) -> (&str, &str, &str, u64) {
+        let s = "UNINIT_PICKLED_STRING";
+        (s, s, s, 0)
+    }
+
+    /// Unpickle from tuple of arguments.
+    fn __setstate__(&mut self, state: &PyAny) -> PyResult<()> {
+        if let Ok(("KafkaInputConfig", brokers, group_id, topics, batch_size)) = state.extract() {
+            self.brokers = brokers;
+            self.group_id = group_id;
+            self.topics = topics;
+            self.batch_size = batch_size;
+            Ok(())
+        } else {
+            Err(PyValueError::new_err(format!(
+                "bad pickle contents for KafkaInputConfig: {state:?}"
+            )))
+        }
+    }
+}
+
+/// Use a user-defined generator as the input source.
+///
+///
+/// Args:
+///
+///     generator: An input_builder function that yields `AdvanceTo()` or `Emit()`
+///         with this worker's input. Must resume from the epoch specified.
+///
+/// Returns:
+///
+///     Config object. Pass this as the `input_config` argument to
+///     your execution entry point.
+
+#[pyclass(module = "bytewax.inputs", extends = InputConfig)]
+#[pyo3(text_signature = "(generator)")]
+pub(crate) struct ManualInputConfig {
+    generator: Py<PyAny>,
```

Comment:
I think it makes sense to rename this bc of the ambiguity around `generator` as a function vs. `generator` as an object-- maybe just back to `input_builder` bc why not.

I am curious about examples where we are returning iterables that are not generators (in manual configuration scenarios using `spawn_cluster` and `cluster_main`), given the required use of `AdvanceTo` and `Emit` now. Do you have one in mind?

Reply:
I guess the term I'm looking for here is [generator iterator](https://docs.python.org/3/glossary.html#term-generator-iterator). I've been using "generator" to refer to the iterator returned, but it looks like it more appropriately refers to the function. A generator is a subset of all functions, though, and we don't require the input builder to be a generator, just that it's a function that returns an iterable.

I don't think we currently have any examples that use non-generator iterators because all of them are using `run` or `run_main`. But when we drop those entry points, we'll be doing lots of things like this in tests:

```py
def ib(i, n, e):
    return [
        Emit("a"),
        Emit("b"),
        AdvanceTo(2),
        Emit("c"),
    ]
```

Reply:
Yeah thanks for that example. I can see how that maintains the contract-- just trying to imagine a user using anything other than a generator. But how often will they rely on manual input in general? 🤷 

Anyhow, back to `input_builder`

Reply:
Yeah, I think long term this `input_builder` structure is not long for this world.

If you want to get creative, you could make a custom iterator transform that isn't a generator, and it's possible some of these exist:

```py
class WrapEmit:
    def __init__(self, wrap):
        self.wrap = wrap

    def next():
        return Emit(self.wrap.next())
```

Actually I'm pretty sure `map(wrap, lambda x: Emit(x))` doesn't result in a generator iterator actually.

```py
In [1]: wrap = [1, 2, 3]

In [2]: from bytewax import Emit

In [4]: inp = map(lambda x: Emit(x), wrap)

In [5]: inp
Out[5]: <map at 0x7f7331b48a60>
```

Looks like it's a `map` object.

None of that really matters, though.